ports/atmel-samd/common-hal/rotaryio/IncrementalEncoder.c

```diff
@@ -0,0 +1,129 @@
+    const mcu_pin_obj_t* pin_a, const mcu_pin_obj_t* pin_b) {
+    if (!pin_a->has_extint || !pin_a->has_extint) {
+        mp_raise_RuntimeError("Both pins must support hardware interrupts");
+    }
+
+    // TODO: The SAMD51 has a peripheral dedicated to quadrature encoder debugging. Use it instead
+    // of the external interrupt.
+
+    if (eic_get_enable()) {
+        if (!eic_channel_free(pin_a->extint_channel) || !eic_channel_free(pin_b->extint_channel)) {
+            mp_raise_RuntimeError("A hardware interrupt channel is already in use");
+        }
+    } else {
+        turn_on_external_interrupt_controller();
+    }
+
+    // These default settings apply when the EIC isn't yet enabled.
+    self->eic_channel_a = pin_a->extint_channel;
+    self->eic_channel_b = pin_b->extint_channel;
+    self->pin_a = pin_a->pin;
+    self->pin_b = pin_b->pin;
+
+    gpio_set_pin_function(self->pin_a, GPIO_PIN_FUNCTION_A);
+    gpio_set_pin_pull_mode(self->pin_a, GPIO_PULL_UP);
+
+    gpio_set_pin_function(self->pin_b, GPIO_PIN_FUNCTION_A);
+    gpio_set_pin_pull_mode(self->pin_b, GPIO_PULL_UP);
+
+    set_eic_channel_data(self->eic_channel_a, (void*) self);
+    set_eic_channel_data(self->eic_channel_b, (void*) self);
+
+    bool pin_a_level = gpio_get_pin_level(self->pin_a);
+    bool pin_b_level = gpio_get_pin_level(self->pin_b);
+    if (!pin_a_level && !pin_b_level) {
+        self->last_state = 1;
+    } else if (!pin_a_level && pin_b_level) {
+        self->last_state = 2;
+    } else if (pin_a_level && pin_b_level) {
+        self->last_state = 3;
+    } else {
+        self->last_state = 4;
+    }
+
+    turn_on_eic_channel(self->eic_channel_a, EIC_CONFIG_SENSE0_BOTH_Val, EIC_HANDLER_INCREMENTAL_ENCODER);
+    turn_on_eic_channel(self->eic_channel_b, EIC_CONFIG_SENSE0_BOTH_Val, EIC_HANDLER_INCREMENTAL_ENCODER);
+}
+
+bool common_hal_rotaryio_incrementalencoder_deinited(rotaryio_incrementalencoder_obj_t* self) {
+    return self->pin_a == NO_PIN;
+}
+
+void common_hal_rotaryio_incrementalencoder_deinit(rotaryio_incrementalencoder_obj_t* self) {
+    if (common_hal_rotaryio_incrementalencoder_deinited(self)) {
+        return;
+    }
+    turn_off_eic_channel(self->eic_channel_a);
+    turn_off_eic_channel(self->eic_channel_b);
+    reset_pin(self->pin_a);
+    self->pin_a = NO_PIN;
+    reset_pin(self->pin_b);
+    self->pin_b = NO_PIN;
+}
+
+mp_int_t common_hal_rotaryio_incrementalencoder_get_position(rotaryio_incrementalencoder_obj_t* self) {
+    return self->position;
+}
+
+void incrementalencoder_interrupt_handler(uint8_t channel) {
+    rotaryio_incrementalencoder_obj_t* self = get_eic_channel_data(channel);
+    // TODO(tannewt): If we need more speed then read the pin directly. gpio_get_pin_level has
+    // smarts to compensate for pin direction we don't need.
+    bool pin_a = gpio_get_pin_level(self->pin_a);
+    bool pin_b = gpio_get_pin_level(self->pin_b);
+
+    uint8_t this_state;
+    if (!pin_a && !pin_b) {
```

Comment:
I think the current algorithm can still jitter the position count due to bounces. For example, consider ab = 00 going to 01. If b bounces it will oscillate between `this_state` being 1 and 2, which will cause `self->position` to be incremented and decremented on each bounce.

There are actually 16 state transitions to consider: ab->a'b', so 2^4 different combinations or states. Some of these are "illegal" and represent going too fast, etc.

I'm writing up a reimplementation that considers all these states. The best implementations I've found online that seems to do the right thing are https://github.com/buxtronix/arduino/blob/master/libraries/Rotary/Rotary.cpp and https://github.com/Erriez/ErriezRotaryEncoderFullStep/blob/master/src/RotaryFullStep.cpp.

The product 377 in the store does a full set of transitions between detents (on each detent both A and B are open).